src/createAction.ts

```diff
@@ -6,51 +6,55 @@ import { Action } from 'redux'
  *
  * @template P The type of the action's payload.
  * @template T the type used for the action type.
+ * @template M The type of the action's meta (optional)
  */
 export type PayloadAction<
   P = any,
   T extends string = string,
   M = void
-> = Action<T> & {
-  payload: P
-} & ([M] extends [void] ? {} : { meta: M })
-
-export type Diff<T, U> = T extends U ? never : T
+  > = WithOptionalMeta<M, WithPayload<P, Action<T>>>;
 
 export type PrepareAction<P> =
   | ((...args: any[]) => { payload: P })
   | ((...args: any[]) => { payload: P; meta: any })
 
+
+export type ActionCreatorWithPreparedPayload<PA extends PrepareAction<any> | void, T extends string = string> =
+  WithTypeProperty<T, PA extends PrepareAction<infer P> ? (...args: Parameters<PA>) => PayloadAction<P, T, MetaOrVoid<PA>> : void>;
+
+export type ActionCreatorWithOptionalPayload<P, T extends string = string> =
+  WithTypeProperty<T, {
+    (payload?: undefined): PayloadAction<undefined, T>
+    <PT extends Diff<P, undefined>>(payload?: PT): PayloadAction<PT, T>
+  }>;
+
+export type ActionCreatorWithoutPayload<T extends string = string> = WithTypeProperty<T, () => PayloadAction<undefined, T>>;
+
+export type ActionCreatorWithPayload<P, T extends string = string> =
+  WithTypeProperty<T, <PT extends P>(payload: PT) => PayloadAction<PT, T>>;
+
 /**
  * An action creator that produces actions with a `payload` attribute.
  */
 export type PayloadActionCreator<
   P = any,
```

Comment:
I could go and rename P to payload, T to Type and PA to PrepareActionFn here, but I'm not really sure if it makes things better or worse in the usages of these below.
Do you have an opinion there?